pkg/clients/sns/snssubscription.go

```diff
@@ -0,0 +1,183 @@
+
+	o := v1alpha1.SNSSubscriptionObservation{}
+	o.Owner = aws.String(attr["Owner"])
+	var status v1alpha1.ConfirmationStatus
+	if s, err := strconv.ParseBool(attr["PendingConfirmation"]); err == nil {
+		if s {
+			status = v1alpha1.ConfirmationPending
+		} else {
+			status = v1alpha1.ConfirmationSuccessful
+		}
+	}
+	o.Status = &status
+
+	if s, err := strconv.ParseBool(attr["ConfirmationWasAuthenticated"]); err == nil {
+		o.ConfirmationWasAuthenticated = aws.Bool(s)
+	}
+
+	return o
+}
+
+// LateInitializeSubscription fills the empty fields in
+// *v1alpha1.SNSSubscriptionParameters with the values seen in
+// sns.Subscription
+func LateInitializeSubscription(in *v1alpha1.SNSSubscriptionParameters, subAttributes map[string]string) {
+	in.DeliveryPolicy = awsclients.LateInitializeStringPtr(in.DeliveryPolicy, aws.String(subAttributes["DeliveryPolicy"]))
+	in.FilterPolicy = awsclients.LateInitializeStringPtr(in.FilterPolicy, aws.String(subAttributes["FilterPolicy"]))
+	in.RawMessageDelivery = awsclients.LateInitializeStringPtr(in.RawMessageDelivery, aws.String(subAttributes["RawMessageDelivery"]))
+	in.RedrivePolicy = awsclients.LateInitializeStringPtr(in.RedrivePolicy, aws.String(subAttributes["RedrivePolicy"]))
+}
+
+// GetSNSSubscription returns SNSSubscription from List of SNSSubscription
+func GetSNSSubscription(res *sns.ListSubscriptionsByTopicResponse, cr *v1alpha1.SNSSubscription) (sns.Subscription, error) {
+
+	for _, sub := range res.Subscriptions {
+		if cmp.Equal(sub.TopicArn, cr.Spec.ForProvider.TopicARN) && cmp.Equal(sub.Endpoint, &cr.Spec.ForProvider.Endpoint) && cmp.Equal(sub.Protocol, &cr.Spec.ForProvider.Protocol) {
+			return sub, nil
+		}
+	}
+
+	return sns.Subscription{}, &SubscriptionNotFound{}
+
+}
+
+// getSubAttributes returns map of SNS Sunscription Attributes
+func getSubAttributes(p v1alpha1.SNSSubscriptionParameters) map[string]string {
+
+	attr := make(map[string]string)
+
+	attr["DeliveryPolicy"] = aws.StringValue(p.DeliveryPolicy)
+	attr["FilterPolicy"] = aws.StringValue(p.FilterPolicy)
+	attr["RawMessageDelivery"] = aws.StringValue(p.RawMessageDelivery)
+	attr["RedrivePolicy"] = aws.StringValue(p.RedrivePolicy)
+
+	return attr
+}
+
+// GetChangedSubAttributes will return the changed attributes  for a subscription
+// in provider side
+func GetChangedSubAttributes(p v1alpha1.SNSSubscriptionParameters, attrs map[string]string) map[string]string {
+	subAttrs := getSubAttributes(p)
+	changedAttrs := make(map[string]string)
+	for k, v := range subAttrs {
+		if v != attrs[k] {
+			changedAttrs[k] = v
+		}
+	}
+
+	return changedAttrs
+}
+
+// IsSNSSubscriptionUpToDate checks if object is up to date
+func IsSNSSubscriptionUpToDate(p v1alpha1.SNSSubscriptionParameters, sub *sns.Subscription, subAttributes map[string]string) bool {
+	return p.Endpoint == aws.StringValue(sub.Endpoint) && p.Protocol == aws.StringValue(sub.Protocol) && isSNSSubscriptionAttributesUpToDate(p, subAttributes)
```

Comment:
Is there much point comparing endpoint and protocol here? I believe they're immutable, so we wouldn't actually update them even if they weren't up to date. In fact, we'd probably enter a tight reconcile loop because we'd continually consider them to need updating, but not update them.